Improve heartbeat

w666 opened the pull request:
Check interval between ping messages more often (interval / 10).
 This is to prevent some configurations that may cause timeouts
 and connection close. For example, when timeout and interval are
 set to very close values and there was data message recived right
 after connection is established.

 Close #255

buhln commented:
I tested the changes and the described failure happens not longer. I did not check the source code in detail.

I tested with the following configuration of the heartbeat:

```
heartbeat: {
        message: 'ping',
        returnMessage: 'pong',
        timeout: 30000,
        interval: 20000,
      }
```

and got the following Dump in wireshark in my local dev env:

![image](https://github.com/user-attachments/assets/322a2e06-f2b0-44ba-81a5-13a5812e09a5)

The black marked packages are authentication and data initalization of the frontend. The blue marked packages are the Ping/Pong heartbeat messages from the backend, controlled by the python `websockets==13.1` package. The red marked packages are the texed based ping/pong heartbeat messages from the frontend.

As visible the first ping messages is send by the fontend after 30 seconds (timeout). Any further ping message is sent after an intervall of 20 seconds. I asume this is the intended behaviour? There are no disconnectes longer as described in the linked issue.

From a functional side the changes are good to go.

robtaussig commented on src/lib/heartbeat.ts:
```diff
@@ -9,31 +9,40 @@ export function heartbeat(ws: WebSocket, lastMessageTime: MutableRefObject<numbe
-    if (lastMessageTime.current + timeout <= Date.now()) {
-      console.warn(`Heartbeat timed out, closing connection, last message was seen ${Date.now() - lastMessageTime.current}ms ago`);
+    const timeNow = Date.now();
+
+    if (lastMessageTime.current + timeout <= timeNow) {
+      console.warn(`Heartbeat timed out, closing connection, last message was seen ${timeNow - lastMessageTime.current}ms ago`);
       ws.close();
     } else {
-      if (lastMessageTime.current + interval <= Date.now()) {
+      if (lastMessageTime.current + interval <= timeNow && lastPingSent + interval <= timeNow) {
         try {
           if (typeof message === 'function') {
             ws.send(message());
           } else {
             ws.send(message);
           }
+          lastPingSent = timeNow;
         } catch (err: unknown) {
           console.error(`Heartbeat failed, closing connection`, err instanceof Error ? err.message : err);
           ws.close();
         }
 
       }
     }
-  }, interval);
+  }, intervalCheck);
```
Correct me if I'm missing the point, but is there a reason to combine these two checks into the same logic? My understanding is that the bug is caused when the interval and timeouts are too close together, but it seems to me that we can avoid some of the confusion by handling their logic separately. For the heartbeat timeout:

```ts
//attach-shared-listeners.ts
const bindMessageHandler = (
  webSocketInstance: WebSocketLike,
  url: string,
  heartbeatOptions?: boolean | HeartbeatOptions
) => {
  let heartbeatTimeoutMs = typeof heartbeatOptions !== 'boolean' ? heartbeatOptions?.timeout : undefined;

  let heartbeatTimeout: NodeJS.Timeout;

  if (heartbeatTimeoutMs) {
    webSocketInstance.addEventListener("close", () => {
      if (heartbeatTimeout) {
        clearTimeout(heartbeatTimeout);
      }
    });
  }

  webSocketInstance.onmessage = (message: WebSocketEventMap['message']) => {
    if (heartbeatTimeoutMs) {
      clearTimeout(heartbeatTimeout);
      heartbeatTimeout = setTimeout(() => {
        webSocketInstance.close();
      }, heartbeatTimeoutMs);
    }

    getSubscribers(url).forEach(subscriber => {
      if (subscriber.optionsRef.current.onMessage) {
        subscriber.optionsRef.current.onMessage(message);
      }

      if (typeof subscriber?.lastMessageTime?.current === 'number') {
        subscriber.lastMessageTime.current = Date.now();
      }

      if (
        typeof subscriber.optionsRef.current.filter === 'function' &&
        subscriber.optionsRef.current.filter(message) !== true
      ) {
        return;
      }

      if (
        heartbeatOptions &&
        typeof heartbeatOptions !== "boolean" &&
        heartbeatOptions?.returnMessage === message.data
      )
        return;

      subscriber.setLastMessage(message);
    });
  };
};
```

And for the interval:

```ts
//attach-shared-listeners.ts
const bindOpenHandler = (
  webSocketInstance: WebSocketLike,
  url: string,
  heartbeatOptions?: boolean | HeartbeatOptions
) => {
  let heartbeatIntervalMs = typeof heartbeatOptions !== 'boolean' ? heartbeatOptions?.interval : undefined;

  let heartbeatInterval: NodeJS.Timeout;

  if (heartbeatIntervalMs) {
    webSocketInstance.addEventListener("close", () => {
      if (heartbeatInterval) {
        clearInterval(heartbeatInterval);
      }
    });
  }

  webSocketInstance.onopen = (event: WebSocketEventMap['open']) => {
    if (heartbeatOptions !== undefined && typeof heartbeatOptions !== 'boolean' && heartbeatIntervalMs && webSocketInstance instanceof WebSocket) {
      heartbeatInterval = setInterval(() => {
        const {
          message = DEFAULT_HEARTBEAT.message
        } = heartbeatOptions;
        webSocketInstance.send(typeof message === 'function' ? message() : message)
      }, heartbeatIntervalMs)
    }
    getSubscribers(url).forEach(subscriber => {
      subscriber.reconnectCount.current = 0;
      if (subscriber.optionsRef.current.onOpen) {
        subscriber.optionsRef.current.onOpen(event);
      }

      subscriber.setReadyState(ReadyState.OPEN);

      let onMessageCb: () => void;

      if (heartbeatOptions && webSocketInstance instanceof WebSocket) {
        subscriber.lastMessageTime.current = Date.now();
        heartbeat(webSocketInstance, subscriber.lastMessageTime, typeof heartbeatOptions === 'boolean' ? undefined : heartbeatOptions,);
      }
    });
  };
};
```

I like having this logic abstracted into its own file, but I kept it in one place here for demonstration purposes.

I also have a question about whether it's better to lift the heartbeat logic outside of the subscriber step (as I did in the above example) -- the way I see it, sharing a single connection between multiple "subscribers" is a front-end implementation detail/optimization. But handling the heartbeat at the subscriber level means that we are sending multiple heartbeat messages for a single websocket connection, which an api might reject/not handle predictably.

w666 replied:
Agree, will refactor this a bit later this week. Thanks for your feedback.

w666 replied:
I refactored it a bit, now it does not create separate interval for each subscriber in shared mode, so only one ping message regardless of the number of subscribers. New test validates this behavior.

Corrected interval in one existing test, but just to make sure it gets expected number of ping messages.

w666 commented:
@buhln,

Could you please retest it? New implementation is more correct, @robtaussig helped to spot one bit I did not think about.

buhln commented:
I get the timout after a while again:

![image](https://github.com/user-attachments/assets/180d011a-14b3-4850-8392-657c416e66b6)

In Wireshark I can see that after this timespan the client stops sending the ping message:

![image](https://github.com/user-attachments/assets/f3756fe2-1803-4078-897e-d4647fb212a7)

1: Last ping/pong messages from the frontend
2: First ping/pong messages after reconnect

The two marked timestamps shows, that after 55s the timout and reconnect from the client is happening:
1373.36s -1318.35s = 55s

So, the issue, that the client stops sending the ping message is still there.

w666 commented:
@buhln 

I added more logging in my last commit. The reason for that is because I can't replicate that problem locally with tests.

Looking on your screenshot from wireshark I noticed that some messages are not ping/pong that are sent as a data, those are real low-level ping-pongs, where `ping` is sent by server and `pong` is sent by client (browser), you can't see those in browser as browsers do not provide access to low-level messages, only to data messages. 

![image](https://github.com/user-attachments/assets/9feee049-200c-401c-9742-8f8adefb1e8f)

the `react-use-websocket` library sends `ping` and receives `pong` messages as text. So, I am not sure about what I see on the screenshot.

I understand that sharing actual network dump is probably not a good idea as it may contain sensitive data. But if you can re-test this again with additional logging that I added in my last commit, then it would provide a bit more data for me.

Thanks.

buhln commented:
You are right. The Ping/Pong messages you are marked are "lowlevel" messages coming from the python websocket packages which runs the backend. The browser replies correctly with a pong.

From my understanding it is visible, that the frontend, so the `react-use-websocket` package is reconnecting due to a timeout of the pong (data/text) message from the backend.

I will test your new version soon and provide the results here. Stay tuned.

buhln commented:
![image](https://github.com/user-attachments/assets/9545e114-2127-430c-b754-46ec6e7d8167)

The first of these disconnects happend after 1222 seconds. Is it possible that this happens if another tab is open in the browser?

w666 commented:
Other tab should not cause any issues.

I assume that Edge browser may put tab into sleep mode if it is not active.
I will try to replicate this tomorrow at work. We use this on a pretty large project.

buhln commented:
I adjusted the timout to your configuration w/o a change in the result. We are using Edge. I think your idea with the sleep mode after some time, espacially if the tab is not active in the browser, is a good theory for the behavior.

In the end it is acceptable because the client connects immediately again as soon as the tab is active again. But the question is, is this the desired behavior?

robtaussig reviewed (commented):
I'm open to your opinion regarding the console.log, but I'm inclined to think that in most cases, it is preferable to not clog up the console with information pertaining to just one (of potentially many) libraries in an app. In an ideal world, we can consider adding a `debug` option to the websocket hook that would allow developers to opt in to this behavior, in which case I'd be all for any and all logs!

vm-orbit commented on src/lib/heartbeat.ts:
```diff
@@ -2,38 +2,55 @@ import { MutableRefObject } from "react";
           } else {
             ws.send(message);
           }
+          console.log(`>>> last message received ${timeNow - lastMessageReceivedAt}ms ago, last ping sent ${timeNow - lastPingSentAt}ms ago`);
```
Hi @robtaussig, @buhln 

I investigated that area, seems like this is known "issue", it is happening from ~2017 in all browsers. Unfortunately, we can't really mitigate the `sleeping tab` problem.

I will remove that debug logging, it was just temporary change to debug it. 

But in this PR in general makes sense, it provides more info when connection is terminated and using just one interval in shared mode.

w666 replied:
Updated PR, removed temp debug logging, it is ready for another review.

buhln replied:
> I investigated that area, seems like this is known "issue", it is happening from ~2017 in all browsers. Unfortunately, we can't really mitigate the `sleeping tab` problem.

Thanks for your research. If this is a "feature" of the browsers you will be right.

But at the end your commit fixed the bug reported initially in https://github.com/robtaussig/react-use-websocket/issues/255. :+1:

robtaussig commented:
@w666 Thank you so much for your time and help with this! Released under [4.12.0](https://www.npmjs.com/package/react-use-websocket/v/4.12.0)
